staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response_test.go

```diff
@@ -222,3 +227,142 @@ func TestWatchEncoderIdentifier(t *testing.T) {
+				}
+				rawListBlueprint, err := base64.StdEncoding.DecodeString(base64ListBlueprint)
+				if err != nil {
+					t.Fatal(err)
+				}
+				if string(rawListBlueprint) != "Use e.passedObject" {
+					t.Fatalf("unexpected rawListBlueprint = %s, expected = Use e.passedObject", rawListBlueprint)
+				}
+				outList := negotiatedEncoder.passedObject.(*v1.PodList)
+				cmp.Equal(outList, makePodList("100"))
+			},
+		},
+		{
+			name:                "encodes the initialEventsListBlueprint as PartialObjectMetadata when requested",
+			targetGVK:           &schema.GroupVersionKind{Group: "meta.k8s.io", Version: "v1", Kind: "PartialObjectMetadata"},
+			actualEvent:         makeBookmarkEventFor(makePodWithInitialEventsAnnotation("2")),
+			actualListBlueprint: makePodList("101"),
+			negotiatedEncoder:   &fakeEncoder{},
+			validate: func(t *testing.T, negotiatedEncoder *fakeEncoder, transformedEvent watch.Event) {
+				if negotiatedEncoder.encodeCounter != 1 {
+					t.Fatalf("expected the rawEncoder to be called exactly 1, not %v", negotiatedEncoder.encodeCounter)
+				}
+				transformedObject := transformedEvent.Object.(*v1.Pod)
+				base64ListBlueprint, ok := transformedObject.Annotations[metav1.InitialEventsListBlueprintAnnotationKey]
+				if !ok {
+					t.Fatalf("the encoded obj doesn't have %q", metav1.InitialEventsListBlueprintAnnotationKey)
+				}
+				rawListBlueprint, err := base64.StdEncoding.DecodeString(base64ListBlueprint)
+				if err != nil {
+					t.Fatal(err)
+				}
+				if string(rawListBlueprint) != "Use e.passedObject" {
+					t.Fatalf("unexpected rawListBlueprint = %s, expected = Use e.passedObject", rawListBlueprint)
+				}
+				outList := negotiatedEncoder.passedObject.(*metav1.PartialObjectMetadataList)
+				expectedList := &metav1.PartialObjectMetadataList{
+					ListMeta: metav1.ListMeta{ResourceVersion: "101"},
+				}
+				cmp.Equal(outList, expectedList)
+			},
+		},
+	}
+
+	for _, scenario := range scenarios {
+		t.Run(scenario.name, func(t *testing.T) {
+			target := newWatchListTransformer(scenario.actualListBlueprint, scenario.targetGVK, scenario.negotiatedEncoder)
+			transformedEvent := target.transform(scenario.actualEvent)
+			scenario.validate(t, scenario.negotiatedEncoder, transformedEvent)
+		})
+	}
+}
+
+func makePodList(rv string) *v1.PodList {
+	return &v1.PodList{
+		ListMeta: metav1.ListMeta{
+			ResourceVersion: rv,
+		},
+		Items: nil,
+	}
+}
+
+func makeBookmarkEventFor(pod *v1.Pod) watch.Event {
+	return watch.Event{
+		Type:   watch.Bookmark,
+		Object: pod,
+	}
+}
+
+func makePod(name string) *v1.Pod {
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   "ns",
+			Annotations: map[string]string{},
+		},
+	}
+}
+
+func makePodWithInitialEventsAnnotation(name string) *v1.Pod {
+	p := makePod(name)
+	p.Annotations[metav1.InitialEventsAnnotationKey] = "true"
+	return p
+}
+
+type fakeEncoder struct {
```

Comment:
I don't understand why we create a fakeEncoder here.
Let's just use an existing regular encoder and operate on a real data that will face in production.

Reply:
From the perspective of this test, we are not interested in the actual encoding. 
We are more interested in which encoder is called and what objects are passed to it. 

Some encoders might actually drop the type information, meaning that decoding to any type could succeed. My intention was to avoid that. For example, encoding an empty PodList to JSON allows it to be decoded into an empty PartialObjectMetadataList. 

In this test, I really want to ensure that a PartialObjectMetadataList was passed and not a PodList.

Reply:
So let's pass PartialObjectMetadaList there and ensure it was properly encoded.

Reply:
Without intercepting encoder there is no way to ensure `PartialObjectMetadaList` was passed.
Which encoder would you like to use ?

What is the benefit of actually encoding the object in a unit test that is not meant to check encoding?
Note that our e2e tests cover the entire path.

Reply:
You just pass appropriate GVK (that you want to encode as PartialObjectMetadataList) and test the whole path.

Reply:
Please be more concrete. Give me the encoder you would like to use and I will write a simple test to check if we can distinguish between `PodList` and `PartialObjectMetadaList `. If the distinction works then we can use it if it doesn't then I would prefer to use the fakeEncoder.

Reply:
You already pass "targetGKV" to the test so that is enough to distinguish what you should receive - if you pass PartialObjectMetadataList (third testcase) you should exactly get it.
So I don't really understand the question here.

Regarding the encoder - the simplest is probably to just use `json.NewSerializer` here, because that gives you a readable output :)

Reply:
ok, used the `json.NewSerializer` but still i would like to assert against the passed obj.

Reply:
Sorry - I think I might still not be clear.
Let me try to write how I envision it:

```
target := newWatchListTransformer(scenario.listBlueprint, scenario.targetGVK, scenario.negotiatedEncoder)
transformedEvent := target.transform(scenario.actualEvent)

actualObjectMeta, err := meta.Accessor(transformedEvent.Object)
if err != nil {
	t.Fatal(err)
}

// code added by me
annotation, ok := actualObjectMeta.GetAnnotations()[metav1.InitialEventsListBlueprintAnnotationKey]
if tc.expectedListBlueprint != "" {
  if annotation != tc.expectedListBlueprint {
    t.Errorf("unexpected annotation...")
  }
} else {
  if ok {
    t.Errorf("unexpected annotation...")
  }
}
```

[and expectedListBlueprint is just a string]

That's all. You don't need any fake encode, because you don't need to check if something was called - if it was, the annotation will be set, if it wasn't, it won't.
And you simply use json serializer as a negotiatedEncoder, because that makes your test human-readable, e.g.
```
  expectedListBlueprint: baseEncodeToString(json.Encode(makePartialObjectMetadataList("101")))
```
[modulo error handling]

Reply:
The thing is that `podList` and `PartialObjectMetadataList` are encoded to the same string. That's why the test checks the passed object. For example:

```
podList: {"metadata":{"resourceVersion":"101"},"items":null}
PartialObjectMetadataList: {"metadata":{"resourceVersion":"101"},"items":null}
```

you can check it with the following code:
```
 e := newFakeEncoder()
 var b bytes.Buffer
 e.Encode(makePodList("101"), &b)
 t.Logf("podList: %v", b.String())
 b.Reset()

 e.Encode(makePartialObjectMetadataList("101"), &b)
 t.Logf("PartialObjectMetadataList: %v", b.String())
```

Reply:
Hmm - what is generally setting TypeMeta? Can we use the encoder that does so to differentiate?

Reply:
We could set the Kind directly in the object, for example:

```
func makePartialObjectMetadataList(rv string) *metav1.PartialObjectMetadataList {
	return &metav1.PartialObjectMetadataList{
		TypeMeta: metav1.TypeMeta{
			Kind: "PartialObjectMetadataList",
		},
		ListMeta: metav1.ListMeta{ResourceVersion: rv},
	}
}

func makePodList(rv string) *v1.PodList {
	return &v1.PodList{
		TypeMeta: metav1.TypeMeta{
			Kind: "PodList",
		},
		ListMeta: metav1.ListMeta{
			ResourceVersion: rv,
		},
		Items: nil,
	}
}
```

then the encoder would produce:
```
podList: {"kind":"PodList","metadata":{"resourceVersion":"101"},"items":null}
PartialObjectMetadataList: {"kind":"PartialObjectMetadataList","metadata":{"resourceVersion":"101"},"items":null}
```

would that be okay with you ?

Reply:
Yeah - I'm fine with that

Reply:
OK, pushed.